### ISSUE TITLE
Cache control for parameters

### ISSUE THREAD

**Issue by kbrandwijk** [apollographql/apollo-cache-control-js#3]
I'm trying to implement cacheControl on a query field that has a filter option:
```
extend type Query {
    getPosts(
      filter: PostFilter
      orderBy: PostOrderBy
      skip: Int
      after: String
      before: String
      first: Int
      last: Int
    ): [Post!]! @cacheControl(maxAge: 240)
  }
```
If I specify a different filter, the results should not be cached. I tried specifying `@cacheControl(maxAge: 0)` on the `filter` field, but the results are still returned from cache.

How can I specify the correct caching, on all of the parameters, so the correct results are returned?

**Comment by martijnwalraven** [comment]
There is currently no way to do this declaratively, but you could do this dynamically from with your resolver. You could look at the `filter` argument there and `cacheControl.setCacheHint({ maxAge: 0 })` depending in the value (see README for a complete code example).

(Sorry for the late response, but I got bitten by the fact that creating a repo doesn't automatically make you watch it.)

**Comment by kbrandwijk** [comment]
Thanks. I'll try out the dynamic solution described in the README. Added complication is that this type definition is actually input for mergeSchemas, so I hope I can do the same thing inside a mergeInfo resolver function...

**Comment by kbrandwijk** [comment]
@martijnwalraven I checked out your proposed approach, but I fail to see how that would solve my issue. I can either set my cacheHint to 0 or a value inside my resolver, so that does not give me an option to make caching dependent of parameters. This looks to me like the cacheControl attribute can only be used on fields without parameters. Specifying a value other than 0 on a field with parameters causes incorrect results, because the cached value is returned, regardless of the parameter values.

To me, honestly, this makes this whole cacheControl thing useless at the moment, because all of my root fields and collections have parameters.

Unless, and I'm just thinking of this while typing, I can add the cacheControl attribute to the Type, instead of the query field, so in my case, setting the cacheControl on the `Post` Type, instead of the `allPosts` query. I'll try that, and report back.

**Comment by martijnwalraven** [comment]
> I can either set my cacheHint to 0 or a value inside my resolver, so that does not give me an option to make caching dependent of parameters. This looks to me like the cacheControl attribute can only be used on fields without parameters. Specifying a value other than 0 on a field with parameters causes incorrect results, because the cached value is returned, regardless of the parameter values.

Could you elaborate on what you mean by this? If you check the `filter` argument from within your `getPosts` resolver, you should be able to use that to set the max age. And if you set it to 0, the result of the query it is a part of will not be cached.

**Comment by kbrandwijk** [comment]
I think my point is that argument values are not part of the caching key (if that's the right way to put it). I want caching to work, but taking argument values into account. Your solution seems to be to just not use caching when an argument is provided.

Let's assume that I use a query on the front page of my app that displays the last ten posts:
` query { allPosts(last: 10) }`. Now, I want that query to be cached for 30 minutes. I can put `cacheControl` on `allPosts`. Now, I manually execute `query { allPosts(last: 5) }`. The result would be the cached 10 posts from my previous query. The alternative would be to not specify `cacheControl` on `allPosts`, but then I loose caching for 98% of the uses of `allPosts`, which is not what I want.

Now, I can introduce a new query: `getLastPosts`, delegate that to `allPosts(last: 10)` and put `cacheControl` on `getLastPosts` and not on `allPosts`, so `query { allPosts(last: 5) }` wouldn't use caching. 

But there are many cases where parameters could be taken into account. For example, having a field with a `format` parameter. I want the results to be cached for each `format`. I can't do that right now.

**Comment by martijnwalraven** [comment]
I think there is a misunderstanding here. The cache key is based on the query hash + variables, meaning argument values are always taken into account. So `query { allPosts(last: 10) }` and `query { allPosts(last: 5) }` are cached separately.

**Comment by kbrandwijk** [comment]
Type definition:
```
extend type Query {
    getPosts(filter: PostFilter
      orderBy: PostOrderBy
      skip: Int
      after: String
      before: String
      first: Int
      last: Int): [Post!]! @cacheControl(maxAge: 240)
  }
```
Resolver:
```
resolvers: mergeInfo => ({
    Query: {
      getPosts: {
        resolve(parent:any, args:any, context:any, info:any) {
          return mergeInfo.delegate('query', 'allPosts', args, context, info)
        },
      },
    })
```
First query:
```
{ getPosts(first: 2) { id title } }
```
Results:
```
{
  "data": {
    "getPosts": [
      {
        "id": "cj978i4u64gq70104vszo84nr",
        "title": "First Post"
      },
      {
        "id": "cj978i9ui4gqe0104xhh1y44z",
        "title": "Second Post"
      }
    ]
  },
  "extensions": {
    "cacheControl": {
      "hints": [
        {
          "maxAge": 240,
          "path": [
            "getPosts"
          ]
        }
      ],
      "version": 1
    }
  }
}
```
Second results for same query (from cache):
```
{
  "data": {
    "getPosts": [
      {
        "id": "cj978i4u64gq70104vszo84nr",
        "title": "First Post"
      },
      {
        "id": "cj978i9ui4gqe0104xhh1y44z",
        "title": "Second Post"
      }
    ]
  }
}
```

Now a query with different parameter values:
```
{ getPosts(first: 1) { id title } }
```
And the results:
```
{
  "data": {
    "getPosts": [
      {
        "id": "cj978i4u64gq70104vszo84nr",
        "title": "First Post"
      },
      {
        "id": "cj978i9ui4gqe0104xhh1y44z",
        "title": "Second Post"
      }
    ]
  }
}
```

**Comment by martijnwalraven** [comment]
I just confirmed this is a bug in the Engine proxy. Thanks for reporting! We should have a fix out for this soon.

Literal values are normalised before computing the cache key, which shouldn't happen. (Variables are taken into account, so the above example would work if you passed in values for `$first` as a variable.)

**Comment by kbrandwijk** [comment]
Great that we got to the bottom of this in the end. I am sorry for any confusing caused in the wording of the initial issue description.

**Comment by martijnwalraven** [comment]
No problem, I'm glad you helped us find this issue!

**Comment by kbrandwijk** [comment]
Now, I'm just hoping that https://github.com/apollographql/graphql-tools/issues/451 will happen soon, so I don't have to create a new query field just to add `@cacheControl` to an existing one.

**Comment by kbrandwijk** [comment]
And one more question regarding the dynamic setting: I thought I had understood that `info` should be treated as immutable in the resolver functions. So why does `info.cacheControl.setCacheHint(...)` not have side-effects?

**Comment by martijnwalraven** [comment]
`info` seemed like a good place to put additional functions that depend on the current execution state on. `info.cacheControl.setCacheHint` doesn't modify `info` itself, but calls a `CacheControlExtension` instance that keeps its own state.

**Comment by kbrandwijk** [comment]
@martijnwalraven Thank you, that is a very good suggestion. I will use that same pattern for my own HOC methods.

**Comment by kbrandwijk** [comment]
@martijnwalraven Any update on this?

**Comment by martijnwalraven** [comment]
@kbrandwijk As far as I know, this has been fixed in recent versions of Engine.

**Comment by kbrandwijk** [comment]
@martijnwalraven Okay, I will retest it and close if confirmed.

**Comment by evans** [comment]
I'm going to close this for now, since it appears that it's been fixed in the Engine Proxy. If there is still an issue, please open an issue on https://github.com/apollographql/apollo-server
